lib/srv/reexec.go

```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
 			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 		}
-		addr := string(buf[:n])
-		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
-			continue
-		}
 
 		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+			var req networking.Request
+			if err := json.Unmarshal(buf[:n], &req); err != nil {
+				slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+				return
+			}
+			log := slog.With("request", req)
+
+			if fn == 0 {
+				log.ErrorContext(ctx, "Networking request requires a control file.")
+				return
+			}
+
+			controlConn, err := uds.FromFile(fbuf[0])
+			_ = fbuf[0].Close()
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to get a connection from control file.")
+				return
+			}
+			defer controlConn.Close()
+
+			log.Debug("Handling networking request")
+			netFile, err := handleNetworkingRequest(ctx, req)
```

Comment:
This context should get closed by a read on `controlConn`:
```suggestion
			ctx, cancel := context.WithCancel(ctx)
			go func() {
				defer cancel()
				_, _ = controlConn.Read(make([]byte, 1))
			}()
			netFile, err := handleNetworkingRequest(ctx, req)
```